src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -98,17 +105,53 @@ def run_python_test(transitive_hydrated_target, pytest, python_setup):
   requirements_pex_response = yield Get(
     ExecuteProcessResult, ExecuteProcessRequest, requirements_pex_request)
 
-  # Gather sources.
+  source_roots = source_root_config.get_source_roots()
+
+  # Gather sources and adjust for the source root.
   # TODO: make TargetAdaptor return a 'sources' field with an empty snapshot instead of raising to
   # simplify the hasattr() checks here!
+  # TODO: find some way to restore the full source name for the stdout of the Pytest run, so that
+  # we can output the original path, rather than the path without the source root. It appears we
+  # do this in V1 via parsing the Junit XML file and adding the value back? See pytest_run.py's
+  # _get_target_from_test(). This is a low priority.
   all_sources_digests = []
   for maybe_source_target in all_targets:
     if hasattr(maybe_source_target.adaptor, 'sources'):
       sources_snapshot = maybe_source_target.adaptor.sources.snapshot
-      all_sources_digests.append(sources_snapshot.directory_digest)
+      digest_adjusted_for_source_root = yield Get(
+        Digest,
+        DirectoryWithPrefixToStrip(
+          sources_snapshot.directory_digest,
+          source_roots.find_by_path(maybe_source_target.adaptor.address.spec_path).path
+        )
+      )
+      all_sources_digests.append(digest_adjusted_for_source_root)
+
+  sources_digest = yield Get(
+    Digest, MergedDirectories(directories=tuple(all_sources_digests)),
+  )
+
+  # TODO: add intrinsic to go from Digest->Snapshot, so that we can avoid having to use
+  # `FileContent`, which unnecessarily gets the content.
+  # TODO: somehow generalize this injection of __init__.py files. Presumably it will be useful
+  # for other Python commands.
+  file_contents = yield Get(FilesContent, Digest, sources_digest)
+  file_paths = [fc.path for fc in file_contents]
+  injected_inits = tuple(sorted(identify_missing_init_files(file_paths)))
+  touch_init_request = ExecuteProcessRequest(
+    argv=("touch",) + injected_inits,
+    output_files=injected_inits,
+    description="Inject empty __init__.py into all packages without one already.",
+    input_files=sources_digest,
+    env={"PATH": os.environ["PATH"]}
```

Comment:
I would really, really like this to be taken out either into its own v2 rule with e.g. a `@rule(FallibleExecuteProcessResult, [LocalPATHExecuteProcessRequest])` which wraps a normal `ExecuteProcessRequest`, does some transformation (setting `env={"PATH": os.environ["PATH"]}`), then does a `yield Get(FallIbleExecuteProcessResult, ExecuteProcessRequest, modified_exe_request)`. I would like this to be done in this PR if possible, as I would prefer to avoid having references to the local PATH start sprouting unchecked (and I would *love* to be able to modify just a single `@rule` in order to modify the behavior of all rules which use anything from the host PATH!).

Reply:
I'm happy to do this in a followup, but request not blocking on this here. 3 V2 Pytest PRs are blocked by this one landing, which block any progress on the remoting project.